Contour levels must be increasing

ianthomas23 opened the pull request:
Fix for issue #5477.  All level arrays passed to contour(f) and tricontour(f) must be in increasing order, otherwise a ValueError is raised.  Added tests and updated documentation.

I have also added extra checks in the C++ contourf and tricontourf generators in case anyone is using the contour generators directly without our python wrappers.

I have not updated api_changes as I cannot decide if this is an API change or a bug fix.

mdboom commented:
As for the question of bugfix vs. API change: What if we made this a warning when the "legacy" setting is `True`, and an exception when any of the new code paths are used?  That sort of splits it down the middle: backward compatible for those that continue to want the broken behavior (but they'll now be informed that it's broken), and fixing for good the new behavior.

tacaswell commented:
:+1: to warning if in legacy mode (which is in line with @WeatherGod https://github.com/matplotlib/matplotlib/issues/5477#issuecomment-156499216) exception in the new mode.

On the other hand, if that makes the checks at the c++ level hard then skip the warnings (I don't know if those check get hit with the legacy code path).

ianthomas23 commented:
Changed from a ValueError to a warning for legacy mode.  I've done this in the python code rather than the legacy C code as I don't want to touch the latter.  Added an extra test for the warning too.

mdboom commented:
:+1: from me!

WeatherGod commented:
pyplot.py needs to be updated using the boilerplate.py, so that pyplot.py gets the updated docstrings, right?

efiring commented:
@weathergod, the docstrings are taken from Axes at import time, so no need to re-run boilerplate.py

WeatherGod commented:
I swear I got plenty of sleep last night! I got that all mixed up in my
head. We would need to rerun boilerplate.py if the call signature changed,
not that the documentation changed.

On Mon, Nov 16, 2015 at 12:47 PM, Eric Firing <notifications@github.com>
wrote:

> @weathergod <https://github.com/weathergod>, the docstrings are taken
> from Axes at import time, so no need to re-run boilerplate.py
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/5485#issuecomment-157116839>
> .
>

WeatherGod commented:
I restarted the spurious failed test. Everything else looks fine to me. I am a bit hesitant about raising a ValueError for 1.5.1/2.0, but I won't hold this up on that account.

My only other comment is if the contour levels must be strictly increasing or if duplicate levels could still be valid? I could imagine some sort of issue where floating point rounding resulted in a list of contour levels that could evaluate as not strictly increasing.

tacaswell commented:
I am not super worried about adding the value exception as we have already broken user code, now we are doing so nosily.

ianthomas23 commented:
@WeatherGod We do need the levels strictly increasing.  If you try to contourf between two identical levels both the legacy and new code will produce incorrect results.  We'll just have to see if any users experience problems due to rounding errors in their contour levels.

WeatherGod commented:
backported to v1.5.x as 67591138a
backported to v2.0.x as b4eecf1ac1

mdboom commented:
@WeatherGod: Just for future reference -- you only need to backport to the earliest branch.  I forward port everything on a regular basis.  (Though I'm sure this did no harm).

WeatherGod commented:
Ok, understood.
